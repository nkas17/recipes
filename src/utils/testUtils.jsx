/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import { HashRouter } from 'react-router-dom';

import configureStore from '../store/configureStore';

// Render with redux. Redefining render of @testing-library/react
function render(
	ui,
	{ initialState, store = configureStore(initialState), ...renderOptions } = {}
) {
	// eslint-disable-next-line react/prop-types
	function Wrapper({ children }) {
		return (
			<HashRouter>
				<Provider store={store}>{children}</Provider>
			</HashRouter>
		);
	}
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';

export { render };
