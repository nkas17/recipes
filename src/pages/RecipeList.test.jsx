/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { render, screen, fireEvent } from '../utils/testUtils';
import '@testing-library/jest-dom/extend-expect';
import RecipePage, { RecipePageTest } from './RecipeList';

const recipes = [
  {
    id: 'tacos',
    title: 'tacos',
    category: 'beef',
    description: 'Tasty homemade tacos - hard or soft',
    ingredients: ['cheese', 'ground beef', 'tortilla'],
  },
  {
    id: 'quiche',
    title: 'quiche',
    category: 'other',
    description: 'Delicious Spinach Quiche and Ham & Cauliflower quiche',
    ingredients: ['cheese', 'pie-crust', 'egg'],
  },
];

describe('RecipesPage', () => {
  it('should render correctly', () => {
    const tree = shallow(
      <RecipePageTest
        history={{}}
        recipes={recipes}
        authenticated
        isLoadingRecipes={false}
        token="tokenValue"
        actions={{}}
      />,
    );
    expect(toJSON(tree)).toMatchSnapshot();
  });

  it('should render correctly while loading', () => {
    const tree = shallow(
      <RecipePageTest
        history={{}}
        recipes={recipes}
        authenticated
        isLoadingRecipes
        token="tokenValue"
        actions={{}}
      />,
    );
    expect(toJSON(tree)).toMatchSnapshot();
  });

  it('should render correctly while unauthenticated', () => {
    const tree = shallow(
      <RecipePageTest
        history={{}}
        recipes={recipes}
        authenticated={false}
        isLoadingRecipes={false}
        token=""
        actions={{}}
      />,
    );
    expect(toJSON(tree)).toMatchSnapshot();
  });
});

const defaultProps = {
  history: {},
};
const defaultState = {
  categories: ['other', 'poultry'],
  numAjaxCallsInProgress: 0,
  user: {
    authenticated: false,
    token: '',
  },
  recipeStore: {
    recipes,
    isLoading: false,
  },
};

describe('RecipePage as a User', () => {
  it('should render page', () => {
    const props = {
      ...defaultProps,
    };
    const state = {
      ...defaultState,
    };
    render(<RecipePage {...props} />, { initialState: state });

    expect(screen.getAllByRole('heading')[0]).toHaveTextContent('lets eat!');

    expect(screen.getAllByRole('link')[0]).toHaveTextContent('recipes');
    expect(screen.getAllByRole('link')[1]).toHaveTextContent('about');
    expect(screen.getAllByRole('link')[2]).toHaveTextContent('log in');

    expect(screen.getAllByRole('heading')[1]).toHaveTextContent('recipes');

    expect(screen.getAllByRole('link')[3]).toHaveTextContent('quiche');
    expect(screen.getAllByRole('link')[4]).toHaveTextContent('tacos');
  });

  it('should search and display filtered recipes', () => {
    const props = {
      ...defaultProps,
    };
    const state = {
      ...defaultState,
    };

    render(<RecipePage {...props} />, { initialState: state });

    // user hasn't searched yet
    expect(screen.getAllByRole('row').length).toBe(2);

    // user searches by title
    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: 'Qui' },
    });
    expect(screen.getAllByRole('row').length).toBe(1);

    // user clears search box
    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: '' },
    });
    expect(screen.getAllByRole('row').length).toBe(2);

    // user searches by category
    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: 'beeF' },
    });
    expect(screen.getAllByRole('row').length).toBe(1);

    // user clears search box
    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: '' },
    });
    expect(screen.getAllByRole('row').length).toBe(2);

    // user searches by description
    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: 'ham' },
    });
    expect(screen.getAllByRole('row').length).toBe(1);
  });
});
