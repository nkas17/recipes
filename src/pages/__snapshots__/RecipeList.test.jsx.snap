// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipesPage should render correctly 1`] = `
<Fragment>
  <Connect(Header) />
  <article
    className="py-2"
  >
    <div
      className="jumbotron"
    >
      <div
        className="recipes-header"
      >
        <h2>
          recipes
        </h2>
        <hr />
      </div>
      <div
        className="row"
      >
        <div
          className="col-12 my-4"
        >
          <input
            className="form-control"
            name="searchRecipe"
            onChange={[Function]}
            placeholder="Search Recipes"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        className="btn btn-link ml-3"
        onClick={[Function]}
        type="button"
      >
        add
      </button>
      <table
        className="table"
      >
        <tbody>
          <RecipeListRow
            deleteRecipe={[Function]}
            key="tacos"
            recipe={
              Object {
                "category": "beef",
                "description": "Tasty homemade tacos - hard or soft",
                "id": "tacos",
                "ingredients": Array [
                  "cheese",
                  "ground beef",
                  "tortilla",
                ],
                "title": "tacos",
              }
            }
            showDelete={true}
          />
          <RecipeListRow
            deleteRecipe={[Function]}
            key="quiche"
            recipe={
              Object {
                "category": "other",
                "description": "Delicious Spinach Quiche and Ham & Cauliflower quiche",
                "id": "quiche",
                "ingredients": Array [
                  "cheese",
                  "pie-crust",
                  "egg",
                ],
                "title": "quiche",
              }
            }
            showDelete={true}
          />
        </tbody>
      </table>
    </div>
  </article>
</Fragment>
`;

exports[`RecipesPage should render correctly while loading 1`] = `
<Fragment>
  <Connect(Header) />
  <article
    className="py-2"
  >
    <div
      className="jumbotron"
    >
      <div
        className="recipes-header"
      >
        <h2>
          recipes
        </h2>
        <hr />
      </div>
      <div
        className="loader"
      />
    </div>
  </article>
</Fragment>
`;

exports[`RecipesPage should render correctly while unauthenticated 1`] = `
<Fragment>
  <Connect(Header) />
  <article
    className="py-2"
  >
    <div
      className="jumbotron"
    >
      <div
        className="recipes-header"
      >
        <h2>
          recipes
        </h2>
        <hr />
      </div>
      <div
        className="row"
      >
        <div
          className="col-12 my-4"
        >
          <input
            className="form-control"
            name="searchRecipe"
            onChange={[Function]}
            placeholder="Search Recipes"
            type="text"
            value=""
          />
        </div>
      </div>
      <table
        className="table"
      >
        <tbody>
          <RecipeListRow
            deleteRecipe={[Function]}
            key="tacos"
            recipe={
              Object {
                "category": "beef",
                "description": "Tasty homemade tacos - hard or soft",
                "id": "tacos",
                "ingredients": Array [
                  "cheese",
                  "ground beef",
                  "tortilla",
                ],
                "title": "tacos",
              }
            }
            showDelete={false}
          />
          <RecipeListRow
            deleteRecipe={[Function]}
            key="quiche"
            recipe={
              Object {
                "category": "other",
                "description": "Delicious Spinach Quiche and Ham & Cauliflower quiche",
                "id": "quiche",
                "ingredients": Array [
                  "cheese",
                  "pie-crust",
                  "egg",
                ],
                "title": "quiche",
              }
            }
            showDelete={false}
          />
        </tbody>
      </table>
    </div>
  </article>
</Fragment>
`;
